src/hooks/useV3Trade.ts

```diff
@@ -0,0 +1,102 @@
+import { BigNumber } from '@ethersproject/bignumber'
+import { Currency, CurrencyAmount, TradeType } from '@uniswap/sdk-core'
+import { Trade } from '@uniswap/v3-sdk'
+import { useRouterTradeExactIn, useRouterTradeExactOut } from 'state/routing/useRouterTrade'
+import useDebounce from './useDebounce'
+import { useLocalV3TradeExactIn, useLocalV3TradeExactOut } from './useLocalV3Trade'
+
+export enum V3TradeState {
+  LOADING,
+  INVALID,
+  NO_ROUTE_FOUND,
+  VALID,
+  SYNCING,
+}
+
+const shouldUseFallback = (state: V3TradeState) => [V3TradeState.NO_ROUTE_FOUND].includes(state)
+
+/**
+ * Returns the best v3 trade for a desired exact input swap.
+ * Uses optimized routes from the Routing API and falls back to the v3 router.
+ * @param amountIn The amount to swap in
+ * @param currentOut the desired output currency
+ */
+export function useV3TradeExactIn(
+  amountIn?: CurrencyAmount<Currency>,
+  currencyOut?: Currency
+): {
+  state: V3TradeState
+  trade: Trade<Currency, Currency, TradeType.EXACT_INPUT> | null
+  gasPriceWei?: BigNumber
+  gasUseEstimate?: BigNumber
+} {
+  const debouncedAmountIn = useDebounce(amountIn, 250)
+
+  // attempt to use multi-route trade
+  const multiRouteTradeExactIn = useRouterTradeExactIn(debouncedAmountIn, currencyOut)
+
+  // consider trade debouncing when inputs/outputs do not match
+  const debouncing =
```

Comment:
Hm that's a bit of a quirkiness of RTK-Query (https://github.com/reduxjs/redux-toolkit/issues/1339#issue-950885472). I need to check that the data returned by RTK-Query is what the input args are expecting, otherwise treat the query as debouncing. RTK-Query is releasing a workaround I can experiment with (https://github.com/reduxjs/redux-toolkit/pull/1500)

Reply:
Habit